src/palace.pyx

```diff
@@ -367,13 +371,16 @@ cdef class Context:
             shared_ptr[alure.MessageHandler](new CppMessageHandler(self)))
 
     def __enter__(self) -> Context:
+        Context.__context_stack.append(self)
         use_context(self)
         return self
 
     def __exit__(self, exc_type: Optional[Type[BaseException]],
                  exc_val: Optional[BaseException],
                  exc_tb: Optional[TracebackType]) -> Optional[bool]:
-        use_context(None)
+        Context.__context_stack.pop()
+        current = Context.current_context()
+        use_context(current)
```

Comment:
Again, this is very risky when there're multiple threads running with different contexts.  BTW the context manager explanation in class' docstring needs update.

Since we have C++ context cached in `previous`, alure API can be used here, e.g. `alure.Context.make_current(self.previous)`.

Reply:
> `alure.Context.make_current(self.previous).`

That would be the same as `use_context` but without converting to Python object right?

Reply:
Yep.